src/utils/exec.c

```diff
@@ -348,190 +423,128 @@ static gboolean _utils_exec_and_report_progress (const gchar **argv, const BDExt
     g_free (args);
     g_free (msg);
 
-    out_pipe = g_io_channel_unix_new (out_fd);
-    err_pipe = g_io_channel_unix_new (err_fd);
-
-    /* set encoding to NULL so we can read binary */
-    g_io_channel_set_encoding (out_pipe, NULL, NULL);
-    g_io_channel_set_encoding (err_pipe, NULL, NULL);
+    /* set both fds for non-blocking read */
+    flags = fcntl (out_fd, F_GETFL, 0);
+    if (fcntl (out_fd, F_SETFL, flags | O_NONBLOCK))
+        g_warning ("_utils_exec_and_report_progress: Failed to set out_fd non-blocking: %m");
+    flags = fcntl (err_fd, F_GETFL, 0);
+    if (fcntl (err_fd, F_SETFL, flags | O_NONBLOCK))
+        g_warning ("_utils_exec_and_report_progress: Failed to set err_fd non-blocking: %m");
 
     if (input) {
-        in_pipe = g_io_channel_unix_new (in_fd);
-        io_status = g_io_channel_write_chars (in_pipe,
-                                              input,
-                                              -1,
-                                              &bytes_written,
-                                              error);
-        if (io_status != G_IO_STATUS_NORMAL) {
-            if (error && (*error))
-                g_prefix_error (error, "Failed to write to stdin of the process: ");
-            else
-                g_set_error (error, BD_UTILS_EXEC_ERROR, BD_UTILS_EXEC_ERROR_FAILED,
-                             "Failed to write to stdin of the process.");
-
+        buf_written = write (in_fd, input, strlen (input));
+        if (buf_written < (ssize_t) strlen (input)) {
+            g_set_error (error, BD_UTILS_EXEC_ERROR, BD_UTILS_EXEC_ERROR_FAILED,
+                         "Failed to write to stdin of the process: %m");
```

Comment:
`write()` doesn't necessarily have to write all the data at once. This needs a loop making sure all bytes where written or there was an actual error.

Reply:
Good point! Added a cycle and some more checks.